drivers/gpio/gpio_fxl6408.c

```diff
@@ -0,0 +1,445 @@
+
+#define SUPPORTED_CONFIG (GPIO_INPUT|GPIO_OUTPUT|GPIO_PULL_DOWN|GPIO_PULL_UP)
+
+/** Configuration data*/
+struct gpio_fxl6408_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+
+	/** The master I2C device's name */
+	const char *const i2c_master_dev_name;
+
+	/** The slave address of the chip */
+	uint16_t i2c_slave_addr;
+};
+
+/** Runtime driver data */
+struct gpio_fxl6408_drv_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+
+	/** Master I2C device */
+	const struct device *i2c_master;
+
+	struct {
+		uint8_t input;
+		uint8_t output;
+		uint8_t dir;
+		uint8_t pud_en;
+		uint8_t pud_sel;
+	} reg_cache;
+
+	struct k_sem lock;
+};
+
+/**
+ * @brief Read the port of certain register function.
+ *
+ * @param dev Device struct of the FXL6408.
+ * @param reg Register to read.
+ * @param cache Pointer to the cache to be updated after successful read.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int read_port_regs(const struct device *dev, uint8_t reg, uint8_t *cache)
+{
+	const struct gpio_fxl6408_config *const config = dev->config;
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+	const struct device *i2c_master = drv_data->i2c_master;
+	uint16_t i2c_addr = config->i2c_slave_addr;
+	uint8_t port_data = 0;
+	int ret = 0;
+
+	ret = i2c_reg_read_byte(i2c_master, i2c_addr, reg, &port_data);
+	if (ret != 0) {
+		LOG_ERR("error reading register 0x%X (%d)", reg, ret);
+		return ret;
+	}
+	LOG_DBG("Read: REG[0x%X] = 0x%X", reg, *cache);
+	*cache = port_data;
+
+	return ret;
+}
+
+/**
+ *  @brief Write to the port registers of certain register function.
+ *
+ * @param dev Device struct of the FXL6408.
+ * @param reg Register to write into. Possible values:  REG_DEVICE_ID_CTRL,
+ * REG_OUTPUT, REG_DIRECTION, REG_PUD_SEL, REG_PUD_EN, REG_OUTPUT_HIGH_Z and
+ * REG_INPUT_DEFAULT.
+ * @param cache Pointer to the cache to be updated after successful write.
+ * @param value New value to set.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int write_port_regs(const struct device *dev, uint8_t reg,
+			   uint8_t *cache, uint8_t value)
+{
+	const struct gpio_fxl6408_config *const config = dev->config;
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+	const struct device *i2c_master = drv_data->i2c_master;
+	int ret = 0;
+
+	ret = i2c_reg_write_byte(i2c_master, config->i2c_slave_addr, reg,
+					 value);
+	if (ret == 0) {
+		*cache = value;
+	} else {
+		LOG_ERR("error writing to register 0x%X (%d)", reg,
+			ret);
+	}
+	LOG_DBG("Write: REG[0x%X] = 0x%X", reg, *cache);
+
+	return ret;
+}
+
+static inline int update_input_regs(const struct device *dev, uint8_t *buf)
+{
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+	int ret = read_port_regs(dev, REG_INPUT_VALUE,
```

Comment:
compared to the PCA95, i changed 
`int read_port_regs(const struct device *dev, uint8_t reg, uint8_t *cache, uint8_t *buf )`
to 
`int read_port_regs(const struct device *dev, uint8_t reg, uint8_t *cache )`
because cache and buf return the same value and I thought this way would be cleaner. I could be wrong though.

Reply:
Didn't notice that. That is fine :)